### ISSUE TITLE
Feat: Implement the functions in the `Bank` contract.

### ISSUE THREAD

**Issue by sparqet** [keep-starknet-strange/satoru#271]
## Description

Implement the `transferOut` function in the `Bank` contract.

File: `src/bank/bank.cairo`.

Function signature:
```rust
    /// Transfer tokens from this contract to a receiver.
    /// # Arguments
    /// * `token` - The token address to transfer.
    /// * `receiver` - The address of the receiver.
    /// * `amount` - The amount of tokens to transfer.
    fn transfer_out(
                ref self: ContractState,
                token: ContractAddress,
                receiver: ContractAddress,
                amount: u128,
            )
```

---

Implement the `transfer_out_handle_native` function in the `Bank` contract.

File: `src/bank/bank.cairo`.

Function signature:
```rust
    /// Transfer tokens from this contract to a receiver
    /// handles native token transfers as well
    /// # Arguments
    /// * `token` - token the token to transfer
    /// * `amount` - amount the amount to transfer
    /// * `receiver` - receiver the address to transfer to
    /// * `shouldUnwrapNativeToken` - shouldUnwrapNativeToken whether to unwrap the wrapped native token before transferring
    fn transfer_out_handle_native(
        ref self: ContractState,
        token: ContractAddress,
        receiver: ContractAddress,
        amount: u128,
        shouldUnwrapNativeToken: bool
    )
```

---

Implement the `transfer_out_native_token` function in the `Bank` contract.

File: `src/bank/bank.cairo`.

Function signature:
```rust
    /// Transfer native tokens from this contract to a receiver
    /// # Arguments
    /// * `amount` - amount the amount to transfer
    /// * `receiver` - receiver the address to transfer to
    fn transfer_out_native_token(
        ref self: ContractState,
        receiver: ContractAddress,
        amount: u128,
    )
```

---

---

Implement the internal `transfer_out_internal` function in the `Bank` contract.

File: `src/bank/bank.cairo`.

Function signature:
```rust
    /// Transfer tokens from this contract to a receiver
    /// # Arguments
    /// * `token` - token the token to transfer
    /// * `amount` - amount the amount to transfer
    /// * `receiver` - receiver the address to transfer to
    fn transfer_out_internal(
        ref self: ContractState,
        token: ContractAddress,
        receiver: ContractAddress,
        amount: u128,
    )
```

---

---

Implement the internal `transfer_out_native_token_internal` function in the `Bank` contract.

File: `src/bank/bank.cairo`.

Function signature:
```rust
    /// Unwrap wrapped native tokens and transfer the native tokens from
    /// this contract to a receiver
    /// # Arguments
    /// * `token` - token the token to transfer
    /// * `amount` - amount the amount to transfer
    /// * `receiver` - receiver the address to transfer to
    fn transfer_out_native_token_internal(
        ref self: ContractState,
        token: ContractAddress,
        receiver: ContractAddress,
        amount: u128,
    )
```

## Acceptance criteria

- Function is implemented and well commented.
- Associated unit tests (nominal and failure) are implemented and passing.

## Resources
- [transferOut function in GMX repo](https://github.com/gmx-io/gmx-synthetics/blob/77a2ff39f1414a105e8589d622bdb09ac3dd97d8/contracts/bank/Bank.sol#L34)
- [transferOut function in GMX repo](https://github.com/gmx-io/gmx-synthetics/blob/77a2ff39f1414a105e8589d622bdb09ac3dd97d8/contracts/bank/Bank.sol#L50)
- [transferOutNativeToken function in GMX repo](https://github.com/gmx-io/gmx-synthetics/blob/77a2ff39f1414a105e8589d622bdb09ac3dd97d8/contracts/bank/Bank.sol#L72)
- [_transferOut function in GMX repo](https://github.com/gmx-io/gmx-synthetics/blob/77a2ff39f1414a105e8589d622bdb09ac3dd97d8/contracts/bank/Bank.sol#L85)
- [_transferOutNativeToken function in GMX repo](https://github.com/gmx-io/gmx-synthetics/blob/77a2ff39f1414a105e8589d622bdb09ac3dd97d8/contracts/bank/Bank.sol#L105)

**Comment by Darlington02** [comment]
@sparqet, can I start with this?

**Comment by Sk8erboi84** [comment]
I'll assign it to you! 🔥

**Comment by Darlington02** [comment]
Hello @sparqet @Sk8erboi84 the the `transfer_out_internal` and `transfer_native_internal` makes a call to a particular TokenUtils method which is currently not available, would I need to also implement the TokenUtils for this PR?

see here for solidity implementation:
https://github.com/gmx-io/gmx-synthetics/blob/main/contracts/token/TokenUtils.sol

**Comment by Sk8erboi84** [comment]
> Hello @sparqet @Sk8erboi84 the the `transfer_out_internal` and `transfer_native_internal` makes a call to a particular TokenUtils method which is currently not available, would I need to also implement the TokenUtils for this PR?
> 
> see here for solidity implementation: https://github.com/gmx-io/gmx-synthetics/blob/main/contracts/token/TokenUtils.sol

Hello @Darlington02 , someone is already working on it and it will probably be ready at noon.

**Comment by Darlington02** [comment]
alright thanks, should open a PR, once that is ready

**Comment by Sk8erboi84** [comment]
> alright thanks, should open a PR, once that is ready

Great!! I'll message you here when the TokenUtils will get merged into main. 🔥

**Comment by Darlington02** [comment]
Can't find any tests implemented for the bank contracts, will take some time to read up on the working architecture tonight, but I need to add tests for the implemented functions for this PR right?

**Comment by sparqet** [comment]
> Can't find any tests implemented for the bank contracts, will take some time to read up on the working architecture tonight, but I need to add tests for the implemented functions for this PR right? or what's meant by:
> 
> 
> 
> > Associated unit tests (nominal and failure) are implemented and passing.

Yes you need to test every function you implement in the PR 👍

**Comment by Darlington02** [comment]
> > Hello @sparqet @Sk8erboi84 the the `transfer_out_internal` and `transfer_native_internal` makes a call to a particular TokenUtils method which is currently not available, would I need to also implement the TokenUtils for this PR?
> > see here for solidity implementation: https://github.com/gmx-io/gmx-synthetics/blob/main/contracts/token/TokenUtils.sol
> 
> Hello @Darlington02 , someone is already working on it and it will probably be ready at noon.

Hey @Sk8erboi84 is this available yet? 🙂

**Comment by Sk8erboi84** [comment]
> > > Hello @sparqet @Sk8erboi84 the the `transfer_out_internal` and `transfer_native_internal` makes a call to a particular TokenUtils method which is currently not available, would I need to also implement the TokenUtils for this PR?
> > > see here for solidity implementation: https://github.com/gmx-io/gmx-synthetics/blob/main/contracts/token/TokenUtils.sol
> > 
> > 
> > Hello @Darlington02 , someone is already working on it and it will probably be ready at noon.
> 
> Hey @Sk8erboi84 is this available yet? 🙂

Hello @Darlington02 , I just asked the person who was suppose to work on the issue and he told me that he wont be able to finish it today and he didn't do much. If you want you can do it instead and ill add it to your issue. Sorry
